source-dynamodb/backfill.go

```diff
@@ -0,0 +1,322 @@
+		}
+
+		if len(scanned.Items) > 0 {
+			// Adjust RCU usage prediction if we got anything from this request.
+			rcusPerRequest = int(*scanned.ConsumedCapacity.CapacityUnits)
+
+			// Adjust limiter bucket size if needed.
+			if rcusPerRequest > limiter.Burst() {
+				log.WithFields(log.Fields{
+					"table":         t.tableName,
+					"segment":       segment,
+					"previousBurst": limiter.Burst(),
+					"newBurst":      rcusPerRequest,
+				}).Debug("increasing rcu limiter burst size")
+				limiter.SetBurst(rcusPerRequest)
+			}
+		}
+
+		if scanned.LastEvaluatedKey != nil {
+			// There is more data available for this segment, so put it back into the queue.
+			t.activeSegments <- segment
+		} else {
+			log.WithFields(log.Fields{
+				"table":   t.tableName,
+				"segment": segment,
+			}).Info("backfill segment complete")
+		}
+	}
+}
+
+func (c *capture) doScan(
+	ctx context.Context,
+	t *table,
+	scanLimit *int32,
+	limiter *rate.Limiter,
+	rcusPerRequest int,
+	segment int,
+	startKey map[string]types.AttributeValue,
+) (*dynamodb.ScanOutput, error) {
+	var (
+		// The DynamoDB client library does retry internally, but occasionally we need to do
+		// additional retries for throughput exceeded & other unhandled, transient exceptions.
+		maxAttempts = 10
+	)
+
+	for attempt := 0; attempt < maxAttempts; attempt++ {
+		if err := limiter.WaitN(ctx, rcusPerRequest); err != nil {
+			return nil, fmt.Errorf("backfill limiter wait: %w", err)
+		}
+
+		scanned, err := c.client.db.Scan(ctx, &dynamodb.ScanInput{
+			TableName:              aws.String(t.tableName),
+			ConsistentRead:         aws.Bool(true),
+			ExclusiveStartKey:      startKey,
+			Limit:                  scanLimit,
+			ReturnConsumedCapacity: types.ReturnConsumedCapacityTotal,
+			Segment:                aws.Int32(int32(segment)),
+			TotalSegments:          aws.Int32(int32(t.totalBackfillSegments)),
+		})
+		if err != nil {
+			var throughputError *types.ProvisionedThroughputExceededException
+			if errors.As(err, &throughputError) {
+				log.WithFields(log.Fields{
+					"table":   t.tableName,
+					"segment": segment,
+					"attempt": attempt,
+					"err":     err.Error(),
+				}).Info("retrying a throughput exceeded exception")
+				continue
+			}
+
+			// See https://github.com/aws/aws-sdk-go-v2/issues/1825
+			if strings.Contains(err.Error(), "use of closed network connection") {
+				log.WithFields(log.Fields{
+					"table":   t.tableName,
+					"segment": segment,
+					"attempt": attempt,
+					"err":     err.Error(),
+				}).Info("retrying a closed network connection error")
+				continue
+			}
+
+			return nil, fmt.Errorf("scanning table %q segment %d: %w", t.tableName, segment, err)
+		}
+
+		return scanned, nil
+	}
+
+	return nil, fmt.Errorf("failed to scan segment %d of table %q after %d attempts", segment, t.tableName, maxAttempts)
+}
+
+// keyAttributeWrapper makes it possible to serialize a DynamoDB attribute value as JSON for
```

Comment:
another option would be to use FoundationDB tuple, also used for encoding of packed keys b/w the runtime & connectors.

CDC connectors [use it](https://github.com/estuary/connectors/blob/main/sqlcapture/tuples.go) to encode row keys that are persisted into state checkpoints.

Not a deal-breaker, but if it's easy to change i'd recommended FDB tuples for overall consistency across connectors.

Reply:
Updated to use FDB tuples for checkpoint resume keys! It wasn't too difficult and makes things a little more compact.